### PR TITLE
Feature 13 funds

### LINKED ISSUE #13: Funds API
As an API user I would like to have access to funds API

### PR DESCRIPTION
**Pull request by anttileppa**
Resolves #13

### CODE REVIEW

File: src/funds/funds_meta.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        """
+        return next((entry for entry in self.get_fund_values_basic() if fund_id == entry["fund_id"]), None)
+
+    def load_funds(self) -> dict:
+        """Loads fund JSON file
+
+        Returns:
+            dict: JSON object
+        """
+        with open(os.environ["FUND_JSON"]) as json_file:
```
Comment (baherami) [inline src/funds/funds_meta.py:223]: I don't understand this access from design point of view. If these files exist in test directory(tests/data), then they are supposed to be used for tests, but it seems that we are accessing to them from implementation. Maybe they should be in a "common" directory in src? Also we need to check if the "path" is found too.

File: src/funds/funds_meta.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    def get_fund_meta_by_fund_code(self, fund_code: str) -> Optional[FundMeta]:
+        """Returns fund meta entry for given fund code
+
+        Args:
+            fund_code (str): Fund code
+
+        Returns:
+            FundMeta: fund meta
+        """
+        return next((entry for entry in self.get_all_fund_metas() if entry["code"] == fund_code), None)
```
Comment (baherami) [inline src/funds/funds_meta.py:105]: cant see "code" in FundMeta.

File: src/funds/funds_meta.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+            self.data = data
+
+        return self.data
+
+    def translate_fund_meta(self, code: str, fund_json_entry: FundJsonEntry) -> FundMeta:
+        """Translates single JSON file entry to FundMeta entry
+
+        Args:
+            code (str): Fund code
+            entry (FundJsonEntry): JSON entry
```
Comment (baherami) [inline src/funds/funds_meta.py:129]: please update this parameter according to function's parameter.

File: src/funds/funds_meta.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        Returns:
+            dict: JSON object
+        """
+        if not self.fund_options:
+            with open(os.environ["FUND_OPTIONS_JSON"]) as json_file:
+                self.fund_options = json.load(json_file)
+
+        return self.fund_options
+
+    def get_fund_values_basic(self) -> dict:
```
Comment (baherami) [inline src/funds/funds_meta.py:264]: dict -> Optional[List[Dict[str, str]]]

File: src/funds/funds_meta.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+                        basic_value[key.strip()] = value
+                    self.fund_values_basic.append(basic_value)
+
+        return self.fund_values_basic
+
+    def create_id(self, fund_code: str) -> UUID:
+        """Creates UUID from fund code
+
+        Args:
+            name (str): [description]
```
Comment (baherami) [inline src/funds/funds_meta.py:286]: please edit the docstring

File: src/funds/funds_meta.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+                        _3y_change=_3y_change,
+                        _5y_change=_5y_change,
+                        _10y_change=_10y_change,
+                        _15y_change=_15y_change,
+                        _20y_change=_20y_change,
+                        profit_projection=profit_projection,
+                        profit_projection_date=profit_projection_date
+                      )
+
+    def parse_csv_date(self, str: str) -> Optional[date]:
```
Comment (baherami) [inline src/funds/funds_meta.py:178]: please change str to a proper name.

File: src/funds/funds_meta.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+        Returns:
+            Optional[date]: date
+        """
+        if "-" == str:
+            return None
+
+        return datetime.strptime(str, "%d.%m.%Y").date()
+
+    def parse_csv_float(self, str: str) -> Optional[float]:
```
Comment (baherami) [inline src/funds/funds_meta.py:192]: here also parameter name should be different from a type. And in this case("str") we are talking of a builtin name too.

File: src/impl/apis/funds_api.py [hunk 8]
```diff
@@ -1,34 +1,49 @@
...
 
-        ...
-        
     async def list_funds(self,
                          first_result: int,
                          max_results: int,
                          token_bearerAuth: TokenModel
                          ) -> List[Fund]:
-        ...
+        fund_metas = self.fundsMetaController.get_all_fund_metas()
```
Comment (baherami) [inline src/impl/apis/funds_api.py:45]: shouldnt we use first and max result parameters here?

File: src/impl/apis/funds_api.py [hunk 9]
```diff
@@ -1,34 +1,49 @@
 # coding: utf-8
 
-from typing import List
+from typing import List, Optional
 import logging
+import uuid
 
+from fastapi import HTTPException
 from fastapi_utils.cbv import cbv
 from spec.apis.funds_api import FundsApiSpec, router as FundsApiRouter
```
Comment (baherami) [inline src/impl/apis/funds_api.py:9]: FundsApiRouter -> fundsApiRouter, and please do the same with other imports of router( lowecase as lowercase) and please change them where they are used too.

**Review by baherami — CHANGES_REQUESTED** [review]

File: src/funds/funds_meta.py (hunk 1 above)
Comment (anttileppa) [reply src/funds/funds_meta.py:223]: In tests we are using test data, in production there will be production data that will be mounted as an external file

File: src/funds/funds_meta.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+                        basic_value[key.strip()] = value
+                    self.fund_values_basic.append(basic_value)
+
+        return self.fund_values_basic
+
+    def create_id(self, fund_code: str) -> UUID:
+        """Creates UUID from fund code
+
+        Args:
+            name (str): fund code
```
Comment (baherami) [inline src/funds/funds_meta.py:287]: name-> fund_code or change paramater name?

File: src/spec/apis/funds_api.py [hunk 11]
```diff
@@ -142,10 +144,24 @@ async def list_historical_values_spec(
...
-            fundId,
+            self.toUuid(fundId),
             first_result,
             max_results,
             start_date,
             end_date,
             token_bearerAuth
         )
+
+    def toUuid(self, str: str) -> UUID:
```
Comment (baherami) [inline src/spec/apis/funds_api.py:155]: missed this one. please changfe str to a proper paramaeter name. also in docstring.

File: src/spec/apis/system_api.py [hunk 12]
```diff
@@ -51,3 +53,17 @@ async def ping_spec(
         return await self.ping(
             
         )
+
+    def toUuid(self, str: str) -> UUID:
```
Comment (baherami) [inline src/spec/apis/system_api.py:57]: here too.

File: src/spec/apis/funds_api.py [hunk 13]
```diff
@@ -142,10 +144,24 @@ async def list_historical_values_spec(
...
+        """Translates str to UUID
+
+        Args:
+            str (str): str
+
+        Returns:
+            UUID: UUID
+        """
+        try:
+          return UUID(str)
```
Comment (baherami) [inline src/spec/apis/funds_api.py:165]: insert one tab before return

File: src/spec/apis/funds_api.py [hunk 14]
```diff
@@ -142,10 +144,24 @@ async def list_historical_values_spec(
...
+        except ValueError:
+          raise HTTPException(status_code=400, detail="Invalid UUID {str}".format(str = str))
```
Comment (baherami) [inline src/spec/apis/funds_api.py:167]: please insert one tab before raise, also remove extra space for str = str -> str=str

File: src/spec/apis/system_api.py [hunk 15]
```diff
@@ -51,3 +53,17 @@ async def ping_spec(
...
+        """Translates str to UUID
+
+        Args:
+            str (str): str
+
+        Returns:
+            UUID: UUID
+        """
+        try:
```
Comment (baherami) [inline src/spec/apis/system_api.py:66]: please insert missing tabs and remove extra space from str = str ( same as mentioned in earlier comment)

File: src/tests/test_funds_api.py [hunk 16]
```diff
@@ -1,11 +1,112 @@
...
+        assert 12.3 == fund["changeData"]["change15y"]
+        assert 25.5 == fund["changeData"]["change20y"]
+        assert -0.32 == fund["profitProjection"]
+        assert "2021-10-12" == fund["profitProjectionDate"]
+
+    def test_find_fund_invalid_id(self,
+                                  client: TestClient,
+                                  user_1_auth: BearerAuth
+                                  ):
+        for id in invalid_uuids:
```
Comment (baherami) [inline src/tests/test_funds_api.py:61]: id -> invalid_uuid

File: src/tests/test_funds_api.py [hunk 17]
```diff
@@ -1,11 +1,112 @@
...
+            url = "/v1/funds/{fund_id}".format(fund_id=id)
```
Comment (baherami) [inline src/tests/test_funds_api.py:62]: you can use f string here safely.
"/v1/funds/{fund_id}".format(fund_id=id) -> f"/v1/funds/{fund_id}" (ofcourse if you change id -> fund_id earlier in loop definition.

File: src/tests/test_funds_api.py [hunk 18]
```diff
@@ -1,11 +1,112 @@
...
+        """
+
+    def assert_list_count(self,
+                          expected_count: int,
+                          client: TestClient,
+                          auth: BearerAuth,
+                          first_result: int,
+                          max_results: int
+                          ):
+        query = "first_result={first_result}&max_results={max_results}".format(
```
Comment (baherami) [inline src/tests/test_funds_api.py:104]: you can change this one too:
        query = "first_result={first_result}&max_results={max_results}".format(
            first_result=first_result,
            max_results=max_results
        ) 
        ->
               query = f"first_result={first_result}&max_results={max_results}"

File: src/tests/test_funds_api.py [hunk 19]
```diff
@@ -1,11 +1,112 @@
...
+            first_result=first_result,
+            max_results=max_results
+        )
+        url = "/v1/funds?{query}".format(query=query)
```
Comment (baherami) [inline src/tests/test_funds_api.py:108]: same here. basically anywhere but when logging something, the f string is more readable for humans.

File: src/tests/testcontainers/keycloak.py [hunk 20]
```diff
@@ -65,15 +67,24 @@ def get_oidc_token_url(self, realm):
 
         Returns:
             str: Access token URL
-        """        
-        return self.get_keycloak_url() + "/realms/" + realm + "/protocol/openid-connect/token"
+        """
+        template = "{url}/realms/{realm}/protocol/openid-connect/token"
+        url = self.get_keycloak_url()
+        return template.format(url=url, realm=realm)
```
Comment (baherami) [inline src/tests/testcontainers/keycloak.py:73]: same here if you wish.

**Review by baherami — CHANGES_REQUESTED** [review]

**Pull request merged**